src/common/ram/namespace.go

```diff
@@ -0,0 +1,81 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ram
+
+import (
+	"sync"
+)
+
+var (
+	// DefaultParserFactoryManager the default parser manager
+	DefaultParserFactoryManager = NewParserFactoryManager()
+)
+
+// NamespaceParserFactory factory func for
+type NamespaceParserFactory func() NamespaceParser
+
+// NamespaceParser the namespace parser interface
+type NamespaceParser interface {
+	// Match returns true if resource match the parser
+	Match(resource Resource) bool
+	// Parse returns Namespace exact from the resource
+	Parse(resource Resource) (Namespace, error)
+}
+
+// Namespace the namespace interface
+type Namespace interface {
+	// Kind returns the kind of namespace
+	Kind() string
+	// Resource returns new resource for subresources with the namespace
+	Resource(subresources ...Resource) Resource
+	// Identity returns identity attached with namespace
+	Identity() interface{}
+}
+
+// ParserFactoryManager manager for parser factories
+type ParserFactoryManager struct {
```

Comment:
Based on my understanding a `factory` is for creating different implementation of same interface, so why we need a manager to manage different factories while now we only have one implementation?